downstreamadapter/eventcollector/event_collector.go

```diff
@@ -415,6 +419,86 @@ func (c *EventCollector) runDispatchMessage(ctx context.Context, inCh <-chan *me
 	}
 }
 
+func (c *EventCollector) controlCongestion(ctx context.Context) error {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return context.Cause(ctx)
+		case <-ticker.C:
+			messages := c.newCongestionControlMessages()
+			for serverID, m := range messages {
+				msg := messaging.NewSingleTargetMessage(serverID, messaging.EventServiceTopic, m)
+				if err := c.mc.SendCommand(msg); err != nil {
+					log.Warn("send congestion control message failed", zap.Error(err))
+				}
+			}
+		}
+	}
+}
+
+func (c *EventCollector) newCongestionControlMessages() map[node.ID]*event.CongestionControl {
+	// collect all changefeeds' available memory quota
+	availables := make(map[common.ChangeFeedID]uint64)
+	for _, quota := range c.ds.GetMetrics().MemoryControl.AreaMemoryMetrics {
+		changefeedID, ok := c.changefeedIDMap.Load(quota.Area())
+		if !ok {
+			continue
+		}
+		availables[changefeedID.(common.ChangeFeedID)] = uint64(quota.AvailableMemory())
+	}
+	if len(availables) == 0 {
+		return nil
+	}
+
+	// calculate each changefeed's available memory quota for each node
+	// by the proportion of the dispatcher on each node.
+	// this is not accurate, we should also consider each node's workload distribution.
+	proportions := make(map[common.ChangeFeedID]map[node.ID]uint64)
+	c.dispatcherMap.Range(func(k, v interface{}) bool {
+		stat := v.(*dispatcherStat)
+		eventServiceID := stat.connState.getEventServiceID()
+		changefeedID := stat.target.GetChangefeedID()
+
+		holder, ok := proportions[changefeedID]
+		if !ok {
+			holder = make(map[node.ID]uint64, len(availables))
```

Comment:
len(availables)) seems incorrect. When dealing with a large number of changefeeds, this approach could lead to large memory wast.

Reply:
`len(availables))` is the number of changefeeds. But I think what you need is the number of cdc nodes.